duckduckgo/BrowserServicesKit
# Improve Asana integration for failed PR checks

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1203301625297703/1205530774734473/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2036
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1673
What kind of version bump will this require?: N/A
CC: @bwaresiak 

**Description**:
This patch adds a standalone GitHub Action called `asana-pr-failed-checks` to report
PR checks runs failures to Asana. This action is reused by iOS and macOS repository
without code duplication.
The action has 2 modes: create task and close task. First is run when a PR checks run fails
(only after merging the PR to the default branch, as asserted in the PR checks workflow definition).
It creates a task in Asana and populates "Workflow ID" custom field with the PR checks workflow run ID.
If that run is retried and passes (e.g. because of a flaky test), the action is called in "close task" mode.
It then searches relevant Asana project section for a task with the given workflow ID and closes the task.
The action maps GitHub user handles to Asana user IDs and based on the map provided in the JSON file.



**Steps to test this PR**:
See [this Asana task](https://app.asana.com/0/1205237866452338/1205566950918437/f) for an example run of the new integration.

If you want to test it yourself:
1. Revert https://github.com/duckduckgo/BrowserServicesKit/pull/514/commits/bfee4e614b5a3eb9c3b8859bd3fb1c59ab53f053 commit.
2. Push to this branch.
3. Wait for the task to be created (and assigned to you).
4. Go to the workflow and have it rerun failed checks.
5. The workflow will still fail but it will close Asana task, because that's how the https://github.com/duckduckgo/BrowserServicesKit/pull/514/commits/bfee4e614b5a3eb9c3b8859bd3fb1c59ab53f053 commit makes it do.
6. Revert your most recent commit to restore production-ready state.



---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
